MultiDc Feature

ashangit opened the pull request:
This PR will provide MultiDC capability
It will be split in 3 parts:

- Usage of host IP so the host IP will be advertise to all cluster member. Currently the advertise IP for cluster member is the service cluster IP which is only avaiable for internal nodes

- Provide CRD capability to define seeds so we can advertise other DC seeds. It will rely on the same mechanism than internal seeds, which means relying on creation of a specific service with seed label

- Provide a mechanism to define a initCluster. Idea here is to ensure that only one of the DC cluster can bootstrap from local seed. The other DC cluster will wait for the external seed to be available (not relying on internal seeds)

geobeau commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -510,3 +527,15 @@ func stringOrDefault(str, def string) string {
+	if hostNetworking {
+		var ok bool
+		ip, ok = svc.ObjectMeta.Labels[naming.IpLabel]
+		if !ok {
```
OPT, NIT: You can inline the assign and ok:
`if thing, ok := h.(Thing); ok {`

geobeau commented on pkg/controllers/cluster/service.go:
```diff
@@ -43,7 +44,15 @@ func (cc *ClusterReconciler) syncMemberServices(ctx context.Context, c *scyllav1
 		}
 		for _, pod := range podlist.Items {
-			memberService := resource.MemberServiceForPod(&pod, c)
+			oldMemberService := &corev1.Service{}
```
Can you add a comment here to explain a bit why we need "oldMemberService" and why it's "old".

ashangit replied:
Updated to currentMemberService
Usage should be look in resource.MemberServiceForPod(&pod, c, currentMemberService)

geobeau reviewed (changes requested):
I would some basic documentation in docs/source/ to show how it works and help the review process upstream

pgoron commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -510,3 +527,15 @@ func stringOrDefault(str, def string) string {
 	return def
 }
+
+func GetIpFromService(svc *corev1.Service, hostNetworking bool) (string, error) {
```
Instead of passing a bool that need to be computed by caller, I would pass a reference to ScyllaCluster object. That way there won't be any ambiguity on how to resolve this boolean value on caller side.

pgoron commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -60,7 +61,17 @@ func MemberServiceForPod(pod *corev1.Pod, cluster *scyllav1.ScyllaCluster) *core
-	return &corev1.Service{
+
+	// No IP can be  found if in pending state and pod should not be replaced
+	if (svc.ObjectMeta.Labels[naming.IpLabel] == "") && pod.Status.Phase == corev1.PodPending {
```
This check makes sense only if hostNetworking is used. Aren't you missing a condition ?
BTW It makes the behavior of this function a bit difficult to apprehend for caller.

pgoron replied:
fixed in later commit, please squash commits by group of features to ease review.

ashangit replied:
in github when a PR is submit it is usually better to append new commit if there are some change and not to force push to be able to see what has change from one push to another
there are no patch set as provided by gerrit
I will see to squash all them when I will push it upstream

pgoron commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -48,7 +49,7 @@ func HeadlessServiceForCluster(c *scyllav1.ScyllaCluster) *corev1.Service {
 }
 
-func MemberServiceForPod(pod *corev1.Pod, cluster *scyllav1.ScyllaCluster) *corev1.Service {
+func MemberServiceForPod(pod *corev1.Pod, cluster *scyllav1.ScyllaCluster, svc *corev1.Service) (*corev1.Service, error) {
```
s/svc/oldMemberService/ to be consistent with rest of your changes

pgoron commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -60,7 +61,17 @@ func MemberServiceForPod(pod *corev1.Pod, cluster *scyllav1.ScyllaCluster) *core
+
+	// No IP can be  found if in pending state and pod should not be replaced
+	if (svc.ObjectMeta.Labels[naming.IpLabel] == "") && pod.Status.Phase == corev1.PodPending {
+		return nil, fmt.Errorf("pod in pending state and service not initialized")
```
OPT: I find it a bit detrimental to generate errors for a condition that can happen by design (false alerts). Ideally we should simply log info, and skip create/update of the resource but I understand it would involve a bit of refactoring with caller and may make upstream contribution more difficult.

ashangit replied:
The usage of the error is currently not optional on that design.
In fact we rely on the fact that the reconcile request failed and is put back to the event reconcile queue to finally create with appropriate IP.
If the event doesn't failed it will be deleted which means that we will never go back to this service creation and the associated pod will never be able to get appropriate IP to startup from the service.
I will see if I can change the design/behavior and send a WIP review to dicuss

ashangit replied:
Here is the WIP commit that permitts to avoid relying on that "false" error: https://github.com/criteo-forks/scylla-operator/pull/10/commits/decdd2c5fcc44136fa8599dd426fb9cd562b9d8b

pgoron commented on pkg/controllers/sidecar/identity/member.go:
```diff
@@ -65,19 +68,30 @@ func Retrieve(ctx context.Context, name, namespace string, kubeclient kubernetes
 }
 
-func (m *Member) GetSeeds(ctx context.Context, kubeClient kubernetes.Interface) ([]string, error) {
+func (m *Member) GetSeeds(ctx context.Context, kubeClient kubernetes.Interface, hostNetworking bool) ([]string, error) {
```
Same as above I would pass reference to ScyllaCluster object

pgoron replied:
fixed in later commit

pgoron reviewed (commented):
I think you should add a design-doc as a markdown file under doc/ to ease discussion with upstream

pgoron commented on config/operator/crd/bases/scylla.scylladb.com_scyllaclusters.yaml:
```diff
@@ -1438,6 +1438,15 @@ spec:
                   description: PollInterval specifies how often upgrade logic polls on state updates. Increasing this value should lower number of requests sent to apiserver, but it may affect overall time spent during upgrade.
                   type: string
               type: object
+            multiDcCluster:
```
just thinking out loud, instead of talking of multiDcCluster, I'm wondering if remoteCluster or externalCluster wouldn't be more appropriate as "dc" may be a bit too restrictive in term of deployment strategies.

ashangit replied:
I've initially named it remote cluster but then decided to respect the naming of the feature in scylla: https://docs.scylladb.com/operating-scylla/procedures/cluster-management/create_cluster_multidc/ which I still think is better

pgoron replied:
make sense. Thanks for the reference.

pgoron commented on pkg/controllers/sidecar/identity/member.go:
```diff
@@ -91,12 +91,19 @@ func Retrieve(ctx context.Context, name, namespace string, kubeclient kubernetes
 }
 
-func (m *Member) GetSeeds(ctx context.Context, kubeClient kubernetes.Interface, hostNetworking bool) ([]string, error) {
+func (m *Member) GetSeeds(ctx context.Context, kubeClient kubernetes.Interface, cluster *v1.ScyllaCluster) ([]string, error) {
```
:-)

ashangit commented:
> I think you should add a design-doc as a markdown file under doc/ to ease discussion with upstream

doc doesn't seem to contain any design docs.
Once we will be happy with this change I will start a discussion on the scylla slack to present this feature and to see if they will need a design doc

pgoron commented on helm/scylla/templates/cluster.yaml:
```diff
@@ -16,6 +16,14 @@ spec:
   {{- end }}
   developerMode: {{ .Values.developerMode }}
   cpuset: {{ .Values.cpuset }}
+  {{- if .Values.multiDcCluster.enable }}
```
.enable**d** ? (see alternator.enabled just above)

pgoron commented on helm/scylla/values.yaml:
```diff
@@ -39,6 +39,14 @@ automaticOrphanedNodeCleanup: false
 
+# Activate multi DC cluster and define external seeds
+multiDcCluster:
+  enable: false
```
enabled (see previous comment)

pgoron commented on pkg/controllers/cluster/cleanup.go:
```diff
@@ -112,6 +125,31 @@ func (cc *ClusterReconciler) orphanedCleanup(ctx context.Context, c *scyllav1.Sc
 }
 
+func (cc *ClusterReconciler) multiDcSeedsCleanup(ctx context.Context, c *scyllav1.ScyllaCluster, svcs *corev1.ServiceList) error {
+	multiDcSeedsCount := int32(len(c.Spec.MultiDcCluster.Seeds))
```
just curious, why int32 conversion is required here ?

ashangit replied:
to be able to do comparison svcIndex > maxIndex

pgoron commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -48,7 +49,7 @@ func HeadlessServiceForCluster(c *scyllav1.ScyllaCluster) *corev1.Service {
 }
 
-func MemberServiceForPod(pod *corev1.Pod, cluster *scyllav1.ScyllaCluster) *corev1.Service {
+func MemberServiceForPod(pod *corev1.Pod, cluster *scyllav1.ScyllaCluster) (*corev1.Service, error) {
```
error doesn't seem required anymore

pgoron commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -60,7 +61,14 @@ func MemberServiceForPod(pod *corev1.Pod, cluster *scyllav1.ScyllaCluster) *core
 	}
-	return &corev1.Service{
+
+	if cluster.Spec.Network.HostNetworking {
```
I would add a small comment here to explain that when host networking is used, pod ip is known only when pod is started.

pgoron commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -74,6 +82,34 @@ func MemberServiceForPod(pod *corev1.Pod, cluster *scyllav1.ScyllaCluster) *core
 	}
+
+	if cluster.Spec.Network.HostNetworking {
+		service.Spec.ClusterIP = corev1.ClusterIPNone
```
Why do you need to reset service clusterIp when using host networking ? I was under impression your change could co-exist with clusterIp

ashangit replied:
It is more to avoid confusion of IP used and to ensure no bug in some release that would leads to use that ClusterIP with host network. As it will be None it will failed and will be easy to identify the issue while with an IP we could start use that IP...

pgoron commented on pkg/controllers/cluster/cluster_controller.go:
```diff
@@ -208,6 +208,42 @@ func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		},
+	}
+
+	err = c.Watch(
```
just clarifying my understanding on how this watch works: each time an event is received on a pod, you identify the owning statefulset and trigger an reconciliation request on the statefulset ? Is this reconciliate request gonna to trigger a reconciliation request on scylla_cluster (via the watch on sts) ?

ashangit replied:
The EnqueueRequestForPod will identify the scylla cluster owner of the sts being the owner of the pod and will emit a reconciliate request for that scylla cluster.

`Is this reconciliate request gonna to trigger a reconciliation request on scylla_cluster (via the watch on sts) ?`
No the reconciliate request doesn't affect other watch. Only real change on the sts will trigger reconciliate request from the watch

pgoron commented on pkg/controllers/cluster/resource/resource.go:
```diff
@@ -510,3 +546,14 @@ func stringOrDefault(str, def string) string {
+	if c.Spec.Network.HostNetworking {
+		var ok bool
+		if ip, ok = svc.ObjectMeta.Labels[naming.IpLabel]; !ok {
+			return "", errors.Errorf("%s label not found on member service %s", naming.IpLabel, svc.Name)
```
I would suggest to add a bit more information in error message like ' (is corresponding pod running ?)'

ashangit replied:
svc.Name is exactly the same than the pod.Name so from this logs it looks quite easy to find the appropriate pod.

pgoron commented on pkg/controllers/cluster/status.go:
```diff
@@ -121,7 +122,11 @@ func (cc *ClusterReconciler) updateStatus(ctx context.Context, cluster *scyllav1
-					rackStatus.ReplaceAddressFirstBoot[svc.Name] = svc.Spec.ClusterIP
+					memberIp, err := resource.GetIpFromService(&svc, cluster)
+					if err != nil {
+						return err
```
I'm wondering if we shouldn't log info/debug and skip this service instead of returning an error. With current code, we won't update other services until pod is running.

ashangit replied:
It doesn't seem fine to me to not throw an issue in that stage
If we reach replacement state for a node it means it should have an IP set so this condition should not happen unless there is a bug or unavailability of k8s
We rely on that status then to be able to replace the pod. Also if we skip it there is a high chance that no other events will lead to reconcile request which means potentially having a pod stuck

pgoron reviewed (commented):
I've reviewed only part concerning host networking (not seed/multi dc stuff).
I'm still a bit concerned by the fact that resource.GetIpFromService can returns error for a nominal case (my concern is for people doing support, seeing errors that are "normal") but I don't see ways to do it without changing too many lines for upstream.
So the remaining question will be which version of your code we should push upstream ? To be discussed.

ashangit commented:
> I've reviewed only part concerning host networking (not seed/multi dc stuff).
> I'm still a bit concerned by the fact that resource.GetIpFromService can returns error for a nominal case (my concern is for people doing support, seeing errors that are "normal") but I don't see ways to do it without changing too many lines for upstream.
> So the remaining question will be which version of your code we should push upstream ? To be discussed.

As indicated in the commit name it is still a WIP one. The idea was in fact more to discuss on the overall implementation to see which one we will push upstream. On my side I have a little preference for that one as relying on error during request reconcile can leads to lots of error logs in the operator and also can delay the action at the end as the re-enqueue request are added in the queue with more and more delay each time it failed to be treated

Commit:
Use PodIP when hostNetwork is enabled instead of member service IP

Currently the advertised IP for cluster member is the IP of the member service.
This IP is not accessible outside of the kubernetes cluster which doesn't permitts to activate multi dc feature.
Member service will provide a new label scylla/IP which will store the PodIP when hostNetwork is enabled.
IP from this label is then used for common operations (replace, scale down) and from the sidecar operator to define listen, seed, rpc address

Commit:
Add multi dc feature for hostNetwork cluster only

Provide capability to enable multi dc feature on scyllacluster crd with the list of distant seeds.
Provide a mechanism to define an initCluster. The init cluster will be the first bootstraped cluster. It will not rely on multi dc seed (only using local ones) so it can bootstrap autinomously.
Other clusters will rely on multi dc seeds only at bootstrap.
When bootstrap is finished meaning all statefulsets are fully created with all members ready a bootstrap status set to finished will be added to the scyllacluster status.
When the finished bootstrap status is reached the cluster will not anymore rely on multi dc seeds (rely on local one only)

Commit:
Add design doc for the multi dc feature
